### PR TITLE
Add Bundle Scanning with EnableSilentMode

### PR DESCRIPTION
**Pull request by eopeter**
This PR seeks to enable Bundle Scanning with Distributed Notification when EnableSilentMode is enabled. In this mode, if a listener is listening to the distributed notification channels `com.google.santa.notification.blockedeexecution.bundlehash.progress` or `com.google.santa.notification.blockedeexecution.bundlehash`, they will be notified of the progress of Bundle Scanning and the hashes produced during the scan on the channels respectively.

This will enable usage where Santa Notification that does not use the Notification Window to be able to still listen for Bundle Scannig Events.

### CODE REVIEW

File: Source/gui/SNTNotificationManager.m [hunk 1]
```diff
@@ -100,7 +100,20 @@ - (void)queueMessage:(SNTMessageWindowController *)pendingMsg {
...
   // If GUI is in silent mode or if there's already a notification queued for
   // this message, don't do anything else.
-  if ([SNTConfigurator configurator].enableSilentMode) return;
+  if ([SNTConfigurator configurator].enableSilentMode) {
+    // hash bundle with path
+    dispatch_async(self.hashBundleBinariesQueue, ^{
+          SNTBinaryMessageWindowController *wc = (SNTBinaryMessageWindowController *)pendingMsg;
+          [self hashBundleBinariesForEvent:wc.event withProgressHandler:^(NSUInteger progressCount) {
+              // progress is reported to the distributed notification
+              [self postDistributedNotification: pendingMsg withProgress: progressCount];
```
Comment (russellhancox) [inline Source/gui/SNTNotificationManager.m:109]: Remove the spaces after the colons in method calls

File: Source/gui/SNTNotificationManager.m [hunk 2]
```diff
@@ -100,7 +100,20 @@ - (void)queueMessage:(SNTMessageWindowController *)pendingMsg {
 
   // If GUI is in silent mode or if there's already a notification queued for
   // this message, don't do anything else.
-  if ([SNTConfigurator configurator].enableSilentMode) return;
+  if ([SNTConfigurator configurator].enableSilentMode) {
+    // hash bundle with path
+    dispatch_async(self.hashBundleBinariesQueue, ^{
+          SNTBinaryMessageWindowController *wc = (SNTBinaryMessageWindowController *)pendingMsg;
```
Comment (russellhancox) [inline Source/gui/SNTNotificationManager.m:106]: The indentation inside this block should be +2 spaces from the previous line (i.e. -4 spaces from where it is now).

Running `./Testing/fix.sh` from your checked out repo should fix this and other style issues automatically.

File: Source/gui/SNTNotificationManager.m [hunk 3]
```diff
@@ -129,6 +142,14 @@ - (void)queueMessage:(SNTMessageWindowController *)pendingMsg {
   }
 }
 
+- (void)postDistributedNotificationWithUserInfo:(NSDictionary *)userInfo notificationName:(NSString *)notificationName {
```
Comment (russellhancox) [inline Source/gui/SNTNotificationManager.m:145]: We enforce a per-line character limit of 100 characters. Break this line before `notificationName:` and then line up the colons of each parameter.

File: Source/gui/SNTNotificationManager.m [hunk 4]
```diff
@@ -170,10 +190,38 @@ - (void)postDistributedNotification:(SNTMessageWindowController *)pendingMsg {
     kSigningChain : signingChain,
   };
 
-  [dc postNotificationName:@"com.google.santa.notification.blockedeexecution"
-                    object:@"com.google.santa"
-                  userInfo:userInfo
-        deliverImmediately:YES];
+  [self postDistributedNotificationWithUserInfo:userInfo
+                                  notificationName:@"com.google.santa.notification.blockedeexecution"];
```
Comment (russellhancox) [inline Source/gui/SNTNotificationManager.m:194]: Line up the colons

File: Source/gui/SNTNotificationManager.m [hunk 5]
```diff
@@ -170,10 +190,38 @@ - (void)postDistributedNotification:(SNTMessageWindowController *)pendingMsg {
...
+}
+
+- (void)postDistributedNotificationWithEvent:(SNTStoredEvent *)event withBundleHash: (NSString *) bundleHash {
```
Comment (russellhancox) [inline Source/gui/SNTNotificationManager.m:197]: No spaces around the second parameters type

File: Source/gui/SNTNotificationManager.m [hunk 6]
```diff
@@ -170,10 +190,38 @@ - (void)postDistributedNotification:(SNTMessageWindowController *)pendingMsg {
...
+  NSDictionary *userInfo = @{
+    kFileSHA256: event.fileSHA256 ?: @"",
+    kFilePath : event.filePath ?: @"",
+    kFileBundleName : event.fileBundleName ?: @"",
+    kFileBundleID : event.fileBundleID ?: @"",
+    kFileBundleHash: bundleHash ?: @"",
+  };
+
+  [self postDistributedNotificationWithUserInfo:userInfo
+                                  notificationName:@"com.google.santa.notification.blockedeexecution.bundlehash"];
```
Comment (russellhancox) [inline Source/gui/SNTNotificationManager.m:207]: Line up colons

File: Source/gui/SNTNotificationManager.m [hunk 7]
```diff
@@ -170,10 +190,38 @@ - (void)postDistributedNotification:(SNTMessageWindowController *)pendingMsg {
...
+}
+
+- (void)postDistributedNotification:(SNTMessageWindowController *)pendingMsg withProgress: (NSUInteger) progressCount {
```
Comment (russellhancox) [inline Source/gui/SNTNotificationManager.m:210]: No spaces around type

File: Source/gui/SNTNotificationManager.m [hunk 8]
```diff
@@ -170,10 +190,38 @@ - (void)postDistributedNotification:(SNTMessageWindowController *)pendingMsg {
...
+  if (![pendingMsg isKindOfClass:[SNTBinaryMessageWindowController class]]) {
+    return;
+  }
+  SNTBinaryMessageWindowController *wc = (SNTBinaryMessageWindowController *)pendingMsg;
+  NSDictionary *userInfo = @{
+    kFileBundleProgress : [NSString stringWithFormat:@"%lu", (unsigned long)progressCount] ?: @"0",
```
Comment (russellhancox) [inline Source/gui/SNTNotificationManager.m:216]: The `?: @"0"` part here isn't needed, I'm pretty sure it's impossible

File: Source/gui/SNTNotificationManager.m [hunk 9]
```diff
@@ -170,10 +190,38 @@ - (void)postDistributedNotification:(SNTMessageWindowController *)pendingMsg {
...
+    kFileSHA256: wc.event.fileSHA256 ?: @"",
+    kFilePath : wc.event.filePath ?: @"",
+    kFileBundleName : wc.event.fileBundleName ?: @"",
+    kFileBundleID : wc.event.fileBundleID ?: @"",
+  };
+
+  [self postDistributedNotificationWithUserInfo:userInfo
+                                  notificationName:@"com.google.santa.notification.blockedeexecution.bundlehash.progress"];
```
Comment (russellhancox) [inline Source/gui/SNTNotificationManager.m:224]: Line up colons

File: Source/gui/SNTNotificationManager.m [hunk 10]
```diff
@@ -190,17 +238,26 @@ - (void)showQueuedWindow {
       if ([self.currentWindowController isKindOfClass:[SNTBinaryMessageWindowController class]]) {
         SNTBinaryMessageWindowController *controller =
           (SNTBinaryMessageWindowController *)self.currentWindowController;
+        controller.foundFileCountLabel.stringValue = @"Searching for files...";  
         dispatch_async(self.hashBundleBinariesQueue, ^{
-          [self hashBundleBinariesForEvent:controller.event withController:controller];
+          [self hashBundleBinariesForEvent:controller.event withProgressHandler:^(NSUInteger progressCount) {
+              dispatch_async(dispatch_get_main_queue(), ^{
```
Comment (russellhancox) [inline Source/gui/SNTNotificationManager.m:244]: Indent -2

File: Source/gui/SNTNotificationManager.m [hunk 11]
```diff
@@ -190,17 +238,26 @@ - (void)showQueuedWindow {
...
+                //  Progress is reported to the root NSProgress
```
Comment (russellhancox) [inline Source/gui/SNTNotificationManager.m:245]: Since space after `//`

File: Source/gui/SNTNotificationManager.m [hunk 12]
```diff
@@ -190,17 +238,26 @@ - (void)showQueuedWindow {
...
+                [controller.progress becomeCurrentWithPendingUnitCount:progressCount];
+              });
+          } completionHandler:^(SNTStoredEvent *event, NSString *bundleHash) {
+              [controller updateBlockNotification:event withBundleHash:bundleHash];
```
Comment (russellhancox) [inline Source/gui/SNTNotificationManager.m:249]: Indent -2

File: Source/gui/SNTNotificationManager.m [hunk 13]
```diff
@@ -190,17 +238,26 @@ - (void)showQueuedWindow {
...
         });
       }
     }
   });
 }
 
 - (void)hashBundleBinariesForEvent:(SNTStoredEvent *)event
-                    withController:(SNTBinaryMessageWindowController *)withController {
-  withController.foundFileCountLabel.stringValue = @"Searching for files...";
+                withProgressHandler:(void (^)(NSUInteger progressCount))progressHandler
```
Comment (russellhancox) [inline Source/gui/SNTNotificationManager.m:259]: Line up colons

File: Source/gui/SNTNotificationManager.m [hunk 14]
```diff
@@ -212,26 +269,24 @@ - (void)hashBundleBinariesForEvent:(SNTStoredEvent *)event
...
+  // Start hashing. Progress is reported 
   [[bc remoteObjectProxy]
     hashBundleBinariesForEvent:event
                          reply:^(NSString *bh, NSArray<SNTStoredEvent *> *events, NSNumber *ms) {
                            // Revert to displaying the blockable event if we fail to calculate the
                            // bundle hash
                            if (!bh)
-                             return [withController updateBlockNotification:event
-                                                             withBundleHash:nil];
+                             return completionHandler(event, nil);
```
Comment (russellhancox) [inline Source/gui/SNTNotificationManager.m:289]: The condition braces can be omitted only for a single line. Either of these are acceptable:

```
if (!bh) return completionHandler(event, nil);
```

```
if (!bh) {
  return completionHandler(event, nil);
}
```

(The previous code here was wrong)

File: Source/gui/SNTNotificationManagerTest.m [hunk 15]
```diff
@@ -23,7 +23,8 @@
 
 @interface SNTNotificationManager (Testing)
 - (void)hashBundleBinariesForEvent:(SNTStoredEvent *)event
-                    withController:(SNTBinaryMessageWindowController *)controller;
+                withProgressHandler:(void (^)(NSUInteger progressCount))progressHandler
```
Comment (russellhancox) [inline Source/gui/SNTNotificationManagerTest.m:26]: Line up colons

**Review by russellhancox — CHANGES_REQUESTED** [review]
Most of these comments are code-style (we follow https://google.github.io/styleguide/objcguide.html pretty strictly). You can probably fix all of them by running `./Testing/fix.sh` in your repo.

**Comment by tburgin** [comment]
Bundle scanning has historically been tied to a GUI pop up. It is I/O and CPU intensive in its current implementation and we want to give the user a way to cancel the scan.

If we were to move forward with silent mode bundle scanning, limits will need to be implemented on the scanning so we don’t hammer the system without and recourse for the user.

**Comment by eopeter** [comment]
> Bundle scanning has been intentionally tied to a GUI pop up. It is I/O and CPU intensive in its current implementation and we want to give the user a way to cancel the scan.
> 
> If we were to move forward with silent mode bundle scanning, limits will need to be implemented on the scanning so we don’t hammer the system without recourse for the user.

Will it be acceptable to add option `EnableSilentModeBundleScanning` so that it has to be set to `true` by the user so they are opting in before we attempt bundle scanning when `EnableSilentMode` is `true`?

File: Source/gui/SNTNotificationManager.m (hunk 1 above)
Comment (eopeter) [reply Source/gui/SNTNotificationManager.m]: fixed with `./Testing/fix.sh`

File: Source/gui/SNTNotificationManager.m (hunk 2 above)
Comment (eopeter) [reply Source/gui/SNTNotificationManager.m]: fixed with `./Testing/fix.sh`

File: Source/gui/SNTNotificationManager.m (hunk 3 above)
Comment (eopeter) [reply Source/gui/SNTNotificationManager.m]: fixed with `./Testing/fix.sh`

File: Source/gui/SNTNotificationManager.m (hunk 4 above)
Comment (eopeter) [reply Source/gui/SNTNotificationManager.m]: fixed with `./Testing/fix.sh`

File: Source/gui/SNTNotificationManager.m (hunk 5 above)
Comment (eopeter) [reply Source/gui/SNTNotificationManager.m]: fixed with `./Testing/fix.sh`

File: Source/gui/SNTNotificationManager.m (hunk 11 above)
Comment (eopeter) [reply Source/gui/SNTNotificationManager.m:254]: fixed with `./Testing/fix.sh`

File: Source/gui/SNTNotificationManager.m (hunk 12 above)
Comment (eopeter) [reply Source/gui/SNTNotificationManager.m:259]: fixed with `./Testing/fix.sh`

File: Source/gui/SNTNotificationManager.m (hunk 13 above)
Comment (eopeter) [reply Source/gui/SNTNotificationManager.m]: fixed with `./Testing/fix.sh`

File: Source/gui/SNTNotificationManager.m (hunk 14 above)
Comment (eopeter) [reply Source/gui/SNTNotificationManager.m]: fixed with `./Testing/fix.sh`

File: Source/gui/SNTNotificationManagerTest.m (hunk 15 above)
Comment (eopeter) [reply Source/gui/SNTNotificationManagerTest.m]: fixed with `./Testing/fix.sh`

File: Source/gui/SNTNotificationManager.m (hunk 6 above)
Comment (eopeter) [reply Source/gui/SNTNotificationManager.m]: fixed with ./Testing/fix.sh

File: Source/gui/SNTNotificationManager.m (hunk 7 above)
Comment (eopeter) [reply Source/gui/SNTNotificationManager.m]: fixed with ./Testing/fix.sh

File: Source/gui/SNTNotificationManager.m (hunk 8 above)
Comment (eopeter) [reply Source/gui/SNTNotificationManager.m]: You are right; removed the extraneous code

File: Source/gui/SNTNotificationManager.m (hunk 10 above)
Comment (eopeter) [reply Source/gui/SNTNotificationManager.m:253]: fixed with ./Testing/fix.sh

**Comment by tburgin** [comment]
I don’t think EnableSilentModeBundleScanning is the right approach. Bundle scanning in its current state is optimized for speed at all costs due to the assumption that a user is waiting for the results. The resource impact on the system is real and should be carefully considered. Allowing Santa to behave in such a way, even behind a flag, would be misguided I believe.

I propose a second bundle scanning implementation that optimizes for low usage of system resources for use during silent mode. I am happy to consult more on how to accomplish this.

**Comment by eopeter** [comment]
The implementation in this PR keeps the existing functionality as is but needs a way for the distributed listener to be able to cancel the scanning. That is the bone of contention here. Am I right? 

@russellhancox, I can attempt to impl your solution as proposed but will this current PR do for now if we can find a way to cancel the scan? 

@tburgin what is your proposal to optimize in SilentMode?

**Comment by russellhancox** [comment]
> needs a way for the distributed listener to be able to cancel the scanning. That is the bone of contention here. Am I right?

We don't want a Santa process to be using significant system resources without the user being able to see what it's doing, how much progress it's making, _and_ have a way to stop it.

> I can attempt to impl your solution as proposed but will this current PR do for now if we can find a way to cancel the scan?

The current PR still allows a scan to begin without a UI present and will impact any Santa user using silent mode right now so I would not be comfortable approving this as-is.

**Comment by tburgin** [comment]
> I think [my proposal](https://github.com/google/santa/issues/1212#issuecomment-1787625603) is a suitable approach without the need for new config keys or a new bundle scanning implementation, it continues to tie scanning to a UI popup but allows a custom implementation to be used.
> 
> We cannot enforce that the distributed listener process is presenting a pop-up while scanning occurs (even if it is very strongly encouraged) but at least the scanning is tied to another process on the system.

If we build this we know there are going to be scans without a UI. I think we should do everything we can to mitigate the impact of those scans. A separate scanning implementation for those instances seems reasonable. We can enforce which implementation is used, we control both ends.

**Comment by eopeter** [comment]
Closing this for now and will handle bundle scanning upon receiving the distributed notification on the listener side.

**Comment by tburgin** [comment]
Russell got me up to speed on the usage of silent mode to support third party custom UIs. His [proposal](https://github.com/google/santa/issues/1212#issuecomment-1787625603) makes a lot more sense to me now and should be the path forward. We don't need a second scanning implementation if the linked plan is implemented. Hope that clears some things up!

**Pull request closed without merging**